File: CleverFerretV2/core/common/src/main/java/com/cleverferret/v2/core/common/error/AppError.java
package com.cleverferret.v2.core.common.error;

/**
 * Sealed, transport-agnostic error hierarchy shared by integrations.
 */
public sealed interface AppError permits
        AppError.TransportFailure,
        AppError.ParsingFailure,
        AppError.AuthenticationFailure,
        AppError.AuthorizationFailure,
        AppError.RateLimitFailure,
        AppError.CircuitBreakerOpen,
        AppError.Throttled,
        AppError.ExternalServiceFailure,
        AppError.UnknownFailure {

    String message();

    record TransportFailure(String message, int httpStatus, boolean timeout) implements AppError {}

    record ParsingFailure(String message, String payloadType) implements AppError {}

    record AuthenticationFailure(String message, String provider) implements AppError {}

    record AuthorizationFailure(String message, String provider) implements AppError {}

    record RateLimitFailure(String message, String provider, long retryAfterMillis) implements AppError {}

    record CircuitBreakerOpen(String message, String integration) implements AppError {}

    record Throttled(String message, String integration) implements AppError {}

    record ExternalServiceFailure(String message, String provider, String errorCode) implements AppError {}

    record UnknownFailure(String message, Throwable cause) implements AppError {}
}

File: CleverFerretV2/core/common/src/main/java/com/cleverferret/v2/core/common/error/AppError.kt
package com.cleverferret.v2.core.common.error

sealed interface AppError {
    val message: String

    data class TransportFailure(override val message: String, val httpStatus: Int, val timeout: Boolean) : AppError
    data class ParsingFailure(override val message: String, val payloadType: String) : AppError
    data class AuthenticationFailure(override val message: String, val provider: String) : AppError
    data class AuthorizationFailure(override val message: String, val provider: String) : AppError
    data class RateLimitFailure(override val message: String, val provider: String, val retryAfterMillis: Long) : AppError
    data class CircuitBreakerOpen(override val message: String, val integration: String) : AppError
    data class Throttled(override val message: String, val integration: String) : AppError
    data class ExternalServiceFailure(override val message: String, val provider: String, val errorCode: String) : AppError
    data class UnknownFailure(override val message: String, val cause: Throwable) : AppError
}
